lib/freebl/algcmac.c

```diff
@@ -0,0 +1,327 @@
+                              unsigned int maxOutputLen,
+                              const unsigned char *input,
+                              unsigned int inputLen) {
+    if (ctx->cipherType == CMAC_AES) {
+        unsigned int tmpOutputLen;
+        SECStatus ret = AES_Encrypt(ctx->cipher.aes, output, &tmpOutputLen, maxOutputLen, input, inputLen);
+
+        if (outputLen != NULL) {
+            *outputLen = tmpOutputLen;
+        }
+
+        return ret;
+    }
+
+    return SECFailure;
+}
+
+/* NIST SP.800-38B, 6.1 Subkey Generation */
+static SECStatus cmac_GenerateSubkeys(CMACContext *ctx) {
+    unsigned char *null_block = PORT_ZNewArray(unsigned char, ctx->blockSize);
+    unsigned char *L = PORT_ZNewArray(unsigned char, ctx->blockSize);
+
+    // Step 1: L = AES(key, null_block)
+    if (cmac_Encrypt(ctx, L, NULL, ctx->blockSize, null_block, ctx->blockSize) != SECSuccess) {
+        ctx->k1 = NULL;
+        ctx->k2 = NULL;
+        return SECFailure;
+    }
+
+    // In the following, some effort has been made to pretend to be constant
+    // time. However, the compiler will likely optimize away the first branch's
+    // XOR operation. While leaking two bits of key information likely won't
+    // affect the security of this algorithm, it is something to be aware of.
+
+    // Step 2: If MSB(L) = 0, K1 = L << 1. Else, K1 = (L << 1) ^ R_b.
+    if ((L[0] & 0x80) == 0) {
+        ctx->k1 = left_shift(L, ctx->blockSize);
+        ctx->k1[ctx->blockSize-1] ^= 0x00;
+    } else {
+        // R_b is 0b10000111 == 135 == 0x87
+        ctx->k1 = left_shift(L, ctx->blockSize);
+        ctx->k1[ctx->blockSize-1] ^= 0x87;
+    }
+
+    // Step 3: If MSB(K1) = 0, K2 = K1 << 1. Else, K2 = (K1 <, 1) ^ R_b.
+    if ((ctx->k1[0] & 0x80) == 0) {
+        ctx->k2 = left_shift(ctx->k1, ctx->blockSize);
+        ctx->k2[ctx->blockSize-1] ^= 0x00;
+    } else {
+        ctx->k2 = left_shift(ctx->k1, ctx->blockSize);
+        ctx->k2[ctx->blockSize-1] ^= 0x87;
+    }
+
+    // Any intermediate value in the computation of the subkey shall be secret.
+    PORT_ZFree(null_block, ctx->blockSize);
+    PORT_ZFree(L, ctx->blockSize);
+
+    // Step 4: Return the values.
+    return SECSuccess;
+}
+
+/* NIST SP.800-38B, 6.2 MAC Generation step 6 */
+static SECStatus cmac_UpdateState(CMACContext *ctx) {
+    if (ctx == NULL || ctx->partialIndex != ctx->blockSize) {
+        PORT_SetError(SEC_ERROR_INVALID_ARGS);
+        return SECFailure;
+    }
+
+    // Step 6: C_i = CIPHER(key, C_{i-1} ^ M_i)  for 1 <= i <= n, and
+    //         C_0 is defined as the empty string.
+
+    for (unsigned int index = 0; index < ctx->blockSize; index++) {
+        ctx->partialBlock[index] ^= ctx->lastBlock[index];
+    }
+
+    return cmac_Encrypt(ctx, ctx->lastBlock, NULL, ctx->blockSize,
+                        ctx->partialBlock, ctx->blockSize);
+}
+
+SECStatus CMAC_AESInit(CMACContext *ctx, const unsigned char *key, unsigned int key_len) {
```

Comment:
I'd rather this be CMAC_Init and take an AES parameter. That way we can add new CMAC algorithms more easily. This iteration can just check that it's AES and fail if it's not.

Reply:
ACK, done.